src/v/io/scheduler.cc

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "io/scheduler.h"
+
+#include "io/logger.h"
+#include "io/persistence.h"
+
+#include <seastar/core/coroutine.hh>
+
+namespace experimental::io {
+
+scheduler::scheduler(size_t nofiles) noexcept
+  : nofiles_(nofiles) {}
+
+void scheduler::add_queue(queue* queue) noexcept {
+    queues_.push_back(*queue);
+    queue->io_queue_.start();
+    queue->monitor_ = monitor(queue);
+}
+
+seastar::future<> scheduler::remove_queue(queue* queue) noexcept {
+    queue->sched_hook_.unlink();
+    queue->stop_ = true;
+    queue->sem_.signal();
+    auto monitor = std::exchange(
+      queue->monitor_, seastar::make_ready_future<>());
+    co_await std::move(monitor);
```

Comment:
nit: it feels like this could be easier to read if the gate is used to control the lifetime of the background fiber

Reply:
do you mean use a gate instead?